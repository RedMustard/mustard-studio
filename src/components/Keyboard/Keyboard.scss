@import '../../style/variables.scss';
@import './BlackKey/BlackKey.scss';
@import './WhiteKey/WhiteKey.scss';

.keyboard {
    display: table-row;
    height: $keyboardHeight;
    position: relative;

    & > * {
        user-select: none;
    }
}
