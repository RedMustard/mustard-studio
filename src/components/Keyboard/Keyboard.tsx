import { h } from 'preact';
import { getNoteFrequencyByKeyNumber } from '../../lib/utils/audio';
import { BlackKey } from './BlackKey/BlackKey';
import { WhiteKey } from './WhiteKey/WhiteKey';
import { KeyboardKeyCount, KeyboardKeyColor } from '../../types/types';

interface KeyboardProps {
    audioContext: AudioContext;
    gainNode: GainNode;
}


export const Keyboard = ({ audioContext, gainNode }: KeyboardProps) => {
    let osc: OscillatorNode;
    let osc2: OscillatorNode;

    const whiteKey = (keyNumber: number) => (
        <WhiteKey
            onMouseDown={(e: MouseEvent) => {
                if (e.buttons === 1) {
                    osc = audioContext.createOscillator();
                    osc.frequency.value = getNoteFrequencyByKeyNumber(keyNumber);
                    osc.type = 'triangle';
                    osc.connect(gainNode);
                    osc.start();

                    osc2 = audioContext.createOscillator();
                    osc2.frequency.value = getNoteFrequencyByKeyNumber(keyNumber);
                    osc2.type = 'square';
                    osc2.connect(gainNode);
                    osc2.start();
                    console.log('kn', keyNumber);
                }
            }}
            onMouseOver={(e: MouseEvent) => {
                if (e.buttons === 1) {
                    osc = audioContext.createOscillator();
                    osc.frequency.value = getNoteFrequencyByKeyNumber(keyNumber);
                    osc.type = 'triangle';
                    osc.connect(gainNode);
                    osc.start();

                    osc2 = audioContext.createOscillator();
                    osc2.frequency.value = getNoteFrequencyByKeyNumber(keyNumber);
                    osc2.type = 'square';
                    osc2.connect(gainNode);
                    osc2.start();
                    console.log('kn', keyNumber);
                }
            }}
            onMouseUp={() => {
                osc.stop();
                osc.disconnect();
                osc2.stop();
                osc2.disconnect();
            }}
            onMouseLeave={(e: MouseEvent) => {
                if (e.buttons === 1) {
                    osc.stop();
                    osc.disconnect();
                    osc2.stop();
                    osc2.disconnect();
                }
            }}
        />
    );

    const blackKey = (keyNumber: number) => (
        <BlackKey
            onMouseDown={(e: MouseEvent) => {
                if (e.buttons === 1) {
                    osc = audioContext.createOscillator();
                    osc.frequency.value = getNoteFrequencyByKeyNumber(keyNumber);
                    osc.type = 'triangle';
                    osc.connect(gainNode);
                    osc.start();

                    osc2 = audioContext.createOscillator();
                    osc2.frequency.value = getNoteFrequencyByKeyNumber(keyNumber);
                    osc2.type = 'square';
                    osc2.connect(gainNode);
                    osc2.start();
                    console.log('kn', keyNumber);
                }
            }}
            onMouseOver={(e: MouseEvent) => {
                if (e.buttons === 1) {
                    osc = audioContext.createOscillator();
                    osc.frequency.value = getNoteFrequencyByKeyNumber(keyNumber);
                    osc.type = 'triangle';
                    osc.connect(gainNode);
                    osc.start();

                    osc2 = audioContext.createOscillator();
                    osc2.frequency.value = getNoteFrequencyByKeyNumber(keyNumber);
                    osc2.type = 'square';
                    osc2.connect(gainNode);
                    osc2.start();
                    console.log('kn', keyNumber);
                }
            }}
            onMouseUp={() => {
                osc.stop();
                osc.disconnect();
                osc2.stop();
                osc2.disconnect();
            }}
            onMouseLeave={(e: MouseEvent) => {
                if (e.buttons === 1) {
                    osc.stop();
                    osc.disconnect();
                    osc2.stop();
                    osc2.disconnect();
                }
            }}
        />
    );

    // Todo: test
    const getKeysByOctaveCount = (octaveCount: number, startingKey: number) => {
        const octaveKeyPattern: KeyboardKeyColor[] = ['white', 'black', 'white', 'black', 'white', 'white', 'black', 'white', 'black', 'white', 'black', 'white'];
        const keys: h.JSX.Element[] = [];

        for (let i = 0; i < octaveCount; i++) {
            octaveKeyPattern.forEach((key: KeyboardKeyColor, index: number) => {
                const octaveSteps = 12;
                const keyNumber = index + (i * octaveSteps) + startingKey;
                if (key === 'white') {
                    keys.push(whiteKey(keyNumber));
                } else if (key === 'black') {
                    keys.push(blackKey(keyNumber));
                }
            });
        }

        return keys;
    };

    return (
        <div class="keyboard">
            {whiteKey(1)}
            {blackKey(2)}
            {whiteKey(3)}
            {getKeysByOctaveCount(7, 4)}
            {whiteKey(88)}
        </div>
    );
};
