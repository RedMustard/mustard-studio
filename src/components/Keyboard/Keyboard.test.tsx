import { h } from 'preact';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { AudioContext } from 'standardized-audio-context-mock';
import { Keyboard } from './Keyboard';


const audioContext = new AudioContext();
const gainNode = audioContext.createGain();

describe('<KeyboardKey />', () => {
    const baseProps = {
        audioContext,
        gainNode,
    };

    it('renders with basic props', () => {
        const wrapper = shallow(
            <Keyboard
                {...baseProps}
            />,
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
