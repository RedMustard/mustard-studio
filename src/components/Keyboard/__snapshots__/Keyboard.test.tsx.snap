// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<KeyboardKey /> renders with basic props 1`] = `
<div
  className="keyboard"
>
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
</div>
`;
